refinery/data_set_manager/utils.py

```diff
@@ -1192,3 +1194,94 @@ def get_solr_response_json(node_uuids):
+        uuid=existing_dataset_uuid
+    )
+    associate_datafiles_from_existing_dataset(
+        existing_dataset, revised_investigation
+    )
+    updated_dataset_title = revised_investigation.title
+    existing_dataset.update_investigation(
+        revised_investigation,
+        "Metadata Revision: for {}".format(updated_dataset_title)
+    )
+    data_set_manager.tasks.annotate_nodes(revised_investigation.uuid)
+    existing_dataset.set_title(updated_dataset_title)
+
+
+@transaction.atomic()
+def associate_datafiles_from_existing_dataset(existing_dataset,
+                                              revised_investigation):
+    """
+    Transfer data files from an existing DataSet's FileStoreItems that
+    correspond to data files of the same source in an Investigation's
+    FileStoreItems
+    :param existing_dataset: DataSet instance to transfer datafiles from
+    :param revised_investigation: Investigation instance to transfer data files
+     to
+    """
+    _check_data_set_cleanliness(existing_dataset)
+    _ensure_revised_investigation_references_existing_datafiles(
+        existing_dataset, revised_investigation
+    )
+
+    existing_investigation = existing_dataset.get_investigation()
+    file_store_items_from_existing_dataset = \
+        existing_investigation.get_file_store_items(local_only=True)
+
+    for new_file_store_item in revised_investigation.get_file_store_items():
+        for prior_file_store_item in file_store_items_from_existing_dataset:
+            if prior_file_store_item.source == new_file_store_item.source:
+                new_file_store_item.datafile = prior_file_store_item.datafile
+                new_file_store_item.save()
+                # It's crucial to clear the datafile of the prior
+                # FileStoreItem as well. Otherwise there would be two
+                # references to the same data file which could cause
+                # unintended side-effects
+                prior_file_store_item.datafile = None
+                prior_file_store_item.save()
+
+
+def _check_data_set_cleanliness(data_set):
+    if not data_set.is_clean():
+        raise RuntimeError("DataSet with UUID: {} is not clean (There have "
+                           "been Analyses or Visualizations performed on it) "
+                           "Remove these objects and try again"
+                           .format(data_set.uuid))
+
+
+def _ensure_revised_investigation_references_existing_datafiles(
+        existing_dataset, revised_investigation
+):
+    existing_investigation = existing_dataset.get_investigation()
+    existing_datafile_names, new_datafile_names = (
+        investigation.get_datafile_names(exclude_metadata_file=True) for
+        investigation in [existing_investigation, revised_investigation]
+    )
+    # Ensure that the names of existing datafiles are a subset of the new ones
+    if not set(existing_datafile_names).issubset(set(new_datafile_names)):
```

Comment:
If we assume we have sets at this point, why not do an assert with an operator? Also, is the default str output ok, and the default list str? 
```python
assert existing_datafile_names <= new_datafile_names,
    'Not all data files from {} are referenced "
    'in the revised metadata file. The following are "
    'missing: {}".format(
        existing_dataset, existing_datafile_names - new_datafile_names
    )

Reply:
oh, I see below where this will be sent to the user.